Make feature reset API response more informative

williamrandolph opened the pull request:
Previously, the Reset API returned a list of which feature states it tried to reset, along with success and failure messages. However, the overall return status was always success (`200 OK`), even if an individual feature state failed. (For example, if wildcard deletions were disabled for associated indices #71150.)

This PR changes the Reset API so that it uses different response codes for different situations: `200 OK` if all reset operations succeed, `207 MULTI STATUS` if some operations succeed and others fail, and `500 INTERNAL SERVER ERROR` if all of the reset operations fail. Clients should check the response code from this API.

The API response now includes formatted stack traces when reset operation fails.

- [x] add documentation about the possible 207 response

Fixes #71178

williamrandolph committed:
Reset API should fail fast when there is an error

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

williamrandolph committed:
Use admin credentials with reset API

jaymode commented:
Hmm, I wonder if this is a prime example for using an HTTP 207 as discussed in https://github.com/elastic/elasticsearch/issues/60442? It seems like it to me where the behavior of the API is desirable so we know all of the errors rather than the first and have to play whack a mole. Maybe we should instead fix the consumers of the API to check for failures? What do you think?

williamrandolph commented:
That seems like a good approach here, and I'll see if I can get it working quickly. Otherwise it might be worth making a new issue for it so that we can get some kind of fail-fast behavior into the tests.

williamrandolph committed:
Merge branch 'master' into reset-api-fail-fast

williamrandolph committed:
Improve feature reset status reporting

Previously, the ResetFeatureStateStatus object captured its status in a
String, which meant that if we wanted to know if something succeeded or
failed, we'd have to parse information out of the string. This isn't a
good way of doing things.

I've introduced a SUCCESS/FAILURE enum for status constants, and added a
check for failures in the transport action. Instead of throwing the
first exception that comes back from the resets, we now throw an
exception that shows a complete response with all successes and
failures. We still "fail fast" when we can't reset the features, but now
the user won't have to iterate in order to find all the errors.

The response message could probably be improved from here. I've got some
todos in the code for reminders.

williamrandolph committed:
Merging changes from master

williamrandolph committed:
Merge branch 'master' into reset-api-fail-fast

williamrandolph committed:
Return a 207 if some but not all reset actions fail

williamrandolph committed:
Merge branch 'master' into reset-api-fail-fast

williamrandolph committed:
Serialize stack trace in usable format

williamrandolph commented on server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestResetFeatureStateAction.java:
```diff
@@ -39,6 +41,19 @@ public String getName() {
+            new RestToXContentListener<>(restChannel) {
+                @Override
+                protected RestStatus getStatus(ResetFeatureStateResponse response) {
+                    if (response.hasAllFailures()) {
```
So I've got a 500 if all resets fail, a 207 for mixed responses, and a 200 if all goes well. Does this make sense?

williamrandolph committed:
Merge branch 'master' into reset-api-fail-fast

williamrandolph committed:
Reduce number of iterations over response set

williamrandolph committed:
Merge branch 'master' into reset-api-fail-fast

williamrandolph commented:
I think these test failures have to do with this branch, so I'm looking into them.

williamrandolph committed:
Add optional exception to HLRC reset feature state response

williamrandolph committed:
Add tests for xcontent serialization

williamrandolph committed:
Make method name more specific

williamrandolph committed:
Add a line to the docs about response codes

williamrandolph committed:
Use constructing object parser to handle exception

jaymode commented on client/rest-high-level/src/main/java/org/elasticsearch/client/feature/ResetFeaturesResponse.java:
```diff
@@ -47,24 +48,30 @@ public static ResetFeaturesResponse parse(XContentParser parser) {
     public static class ResetFeatureStateStatus {
         private final String featureName;
         private final String status;
+        private final Exception exception;
```
I cannot comment on the actual line but on line 26, the object parser still uses the "snapshottable" term and is incorrect for this class. Can you address this?

jaymode commented on client/rest-high-level/src/main/java/org/elasticsearch/client/feature/ResetFeaturesResponse.java:
```diff
@@ -78,5 +85,9 @@ public String getFeatureName() {
             return status;
         }
+
+        public Exception getException() {
```
Mark as `@Nullable` and add javadocs to this method and the class?

jaymode commented on client/rest-high-level/src/main/java/org/elasticsearch/client/feature/ResetFeaturesResponse.java:
```diff
@@ -47,24 +48,30 @@ public static ResetFeaturesResponse parse(XContentParser parser) {
         }
 
-        ResetFeatureStateStatus(String featureName, String status) {
+        ResetFeatureStateStatus(String featureName, String status, Exception exception) {
```
```suggestion
        ResetFeatureStateStatus(String featureName, String status, @Nullable Exception exception) {
```

jaymode commented on server/src/internalClusterTest/java/org/elasticsearch/snapshots/FeatureStateResetApiIT.java:
```diff
@@ -94,6 +101,25 @@ public void testResetSystemIndices() throws Exception {
+    /**
+     * Evil test - what if the cleanup method fails?
+     */
+    public void testFailFastOnStateCleanupFailure() throws Exception {
```
technically this is not failing fast since we let everything finish before returning a response?

williamrandolph replied:
Good catch — I've changed the name.

jaymode commented on server/src/internalClusterTest/java/org/elasticsearch/snapshots/FeatureStateResetApiIT.java:
```diff
@@ -94,6 +101,25 @@ public void testResetSystemIndices() throws Exception {
     }
 
+    /**
+     * Evil test - what if the cleanup method fails?
```
```suggestion
     * Evil test - test behavior when a cleanup method fails
```

jaymode commented on server/src/internalClusterTest/java/org/elasticsearch/snapshots/FeatureStateResetApiIT.java:
```diff
@@ -94,6 +101,25 @@ public void testResetSystemIndices() throws Exception {
+     */
+    public void testFailFastOnStateCleanupFailure() throws Exception {
+        try {
+            EvilSystemIndexTestPlugin.beEvil = true;
```
This variable should be modified using a synchronized static method for safety

jaymode commented on server/src/internalClusterTest/java/org/elasticsearch/snapshots/FeatureStateResetApiIT.java:
```diff
@@ -145,4 +171,34 @@ public String getFeatureDescription() {
+        public String getFeatureDescription() {
+            return "a plugin that can be very bad";
+        }
+
```
```suggestion

public static synchronized void setBeEvil(boolean evil) {
    beEvil = evil;
}

public static synchronized boolean isEvil() {
    return beEvil;
}

```

jaymode commented on server/src/internalClusterTest/java/org/elasticsearch/snapshots/FeatureStateResetApiIT.java:
```diff
@@ -145,4 +171,34 @@ public String getFeatureDescription() {
+            ClusterService clusterService,
+            Client client,
+            ActionListener<ResetFeatureStateResponse.ResetFeatureStateStatus> listener) {
+            if (beEvil == false) {
+                listener.onResponse(ResetFeatureStateResponse.ResetFeatureStateStatus.success(getFeatureName()));
+            } else {
+                listener.onResponse(ResetFeatureStateResponse.ResetFeatureStateStatus.failure(getFeatureName(), "problem!"));
+            }
```
```suggestion
            if (isEvil()) {
                listener.onResponse(ResetFeatureStateResponse.ResetFeatureStateStatus.failure(getFeatureName(), "problem!"));
            } else {
                listener.onResponse(ResetFeatureStateResponse.ResetFeatureStateStatus.success(getFeatureName()));
            }
```

jaymode commented on server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/features/ResetFeatureStateResponse.java:
```diff
@@ -43,7 +44,7 @@ public ResetFeatureStateResponse(StreamInput in) throws IOException {
     }
 
-    public List<ResetFeatureStateStatus> getItemList() {
+    public List<ResetFeatureStateStatus> getFeatureStateResetStatusList() {
```
```suggestion
    public List<ResetFeatureStateStatus> getFeatureStateResetStatuses() {
```

williamrandolph replied:
Thanks for catching this. I may read some _Effective Java_ as penance for this mistake.

jaymode commented on server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/features/ResetFeatureStateResponse.java:
```diff
@@ -92,59 +93,106 @@ public String toString() {
+                exception);
+        }
+
+        public static ResetFeatureStateStatus failure(String featureName, String exceptionMessage) {
```
Let's make the caller provide the exception and not allow for a message to be provided without an exception

jaymode commented on server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/features/ResetFeatureStateResponse.java:
```diff
@@ -92,59 +93,106 @@ public String toString() {
             return this.status;
         }
 
+        public Exception getException() {
```
Mark as `@Nullable`

jaymode commented on server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/features/ResetFeatureStateResponse.java:
```diff
@@ -92,59 +93,106 @@ public String toString() {
             builder.startObject();
             builder.field("feature_name", this.featureName);
             builder.field("status", this.status);
+            ElasticsearchException.generateFailureXContent(builder, params, exception, true);
```
Guard this with a null check since there was no error?

jaymode reviewed (changes requested):
Looking pretty good. I left some comments.

williamrandolph committed:
Improve naming in ResetFeaturesResponse and add javadoc

williamrandolph committed:
Use concurrency-safe accessors for static variable

williamrandolph committed:
Respond to PR feedback

williamrandolph committed:
Pity me and my XContent woes

williamrandolph committed:
Merge branch 'master' into reset-api-fail-fast

jaymode commented on client/rest-high-level/src/main/java/org/elasticsearch/client/feature/ResetFeaturesResponse.java:
```diff
@@ -8,21 +8,28 @@
 
+/**
+ * This class represents the response of the Feature State Reset API. It is a
+ * list containing the response of every feature whose state was reset.
```
```suggestion
 * list containing the response of every feature whose state can be reset. The
 * response from each feature will indicate success or failure. In the case of a
 * failure, the cause will be returned as well.
```

jaymode reviewed (approved):
LGTM. Please make sure to backport to 7.x and the 7.13 branches so this makes it into the next 7.13 BC.

williamrandolph committed:
Improve javadoc

Co-authored-by: Jay Modi <jaymode@users.noreply.github.com>

williamrandolph commented:
Backport to 7.x: https://github.com/elastic/elasticsearch/pull/72332
Backport to 7.13: https://github.com/elastic/elasticsearch/pull/72341
